mimxrt: Add Quadrature Encoder and Pulse Counter classes.

robert-hh opened the pull request:
This PR adds Quadrature Encoder and Pulse Counter classes based on the Encoder hardware of the mimxrt MCUs. The base methods are as simple as possible, with options to make use of the hardware features supporting fast encoder sensors, like index pulses.

Tested with a slow manual encoder and a simulation of fast encoder/counter signals up to a input frequency of 25 MHz.

robert-hh committed:
mimxrt: Implement a Quadrature Encoder and Counter class.

These classes are base on the Quadrature Encoder blocks
of the i.MXRT MCUs. The i.MXRT 102x has two encoders, the
other ones four. The i.MXRT 101x does not support this
function. It is implemented as two classes, Encoder and Counter.

The number of pins that can be uses as inputs is limited by
the MCU architecture and the board schematics. The Encoder
class supports:
- Defining the module
- Defining the input pins.
- Defining a pin for an index signal.
- Defining a pin for a Home signal.
- Defining an output pin showing the compare match signal.
- Setting the number of cycles per revolution.
- Setting the initial value for the position.
- Setting the counting direction.
- Setting a glitch filter.
- Setting the value counter as signed or unsigned integers.
- Defining callbacks for getting to a specific position,
  overrun and underrun (starting the next revolution). These
  callbacks can be hard interrupts to ensure short latency.

The encoder counts all phases of a cycle. The span
for the position is 2**32, for the revolution is 2**16. The
highest input frequency is CPU-Clock/24.

The Counter mode counts single pulses on input A of the Encoder.
The configuration support:

- Defining the module
- Defining the input pin.
- Defining the counting direction, either fixed or controlled
  by the level of an input pin.
- Defining a pin for an index signal.
- Defining an ouput pin showing th compare match signal.
- Setting the counter value.
- Setting the glitch filter.
- Returing the value counter as signed or unsigned integer.
- Defining a callback which is called at a certain value.

The counting range is 0 - 2**32-1 and a 16 bit overrun counter.
The highest input frequency is CPU-Clock/12.

robert-hh commented:
Another goodie for the mimxrt port, now at the merge front. Updated and rebased. The documentation for it is in PR #7494.

robert-hh commented:
Since there are now four Quadrature Encoder/Counter classes around, two for ESP32, one for Mimxrt and the generic Python one of @pythoncoder, I made a small comparison table so show similarities and differences. It should be the aim to have at least a common subset of functions and methods. I added the ods file as well for corrections/extensions
![encoder-compare](https://user-images.githubusercontent.com/12476868/144744566-5e590276-fee7-434c-99d7-5652b9ea67bb.png)

[encoder-compare.ods](https://github.com/micropython/micropython/files/7655753/encoder-compare.ods)

peterhinch commented:
@robert-hh There is also [Dave Hylands' offering](https://github.com/dhylands/upy-examples/blob/master/encoder.py). This is an STM specific hardware based solution. From looking at the hardware I think it may be capable of bit-perfect results but I haven't tested this.

robert-hh committed:
mimxrt: Add missing definition for the OLIMEX_RT1010.

robert-hh committed:
mimxrt: Add settings for MIMXRT1015.

The MIMXRT1015 MCU has only one encoder/counter unit.

robert-hh committed:
mimxrt: Add the documentation for the Encoder/Counter class.

This is the MIMXRT specific version, compliant to the
documentation of PR #8072 in the basic methods.

robert-hh commented:
@dpgeorge Now the a documentation is online for the MIMXRT branch, I could add the documentation file for the Encoder/Counter to the PR. Since a while, the PR is ready. You seem to consider how to proceed with the various Encoder/Counter implementations that are around. @IhorNehrutsa, @peterhinch, @jonathanhogg and me agreed on a basic set of API methods, which are extended by the port specific implementations.
There is a open discussion about whether to have a simple hardware driver and a Python shim on top of it, whi extends the machine module, or a full C implementation. The solution of @jonathanhogg uses the Python + C approach, the PR of @IhorNehrutsa  the pure C way. For STM32 a combined C + Python  approach could also be a feasible way. 
About the C + Python method, I have a question. Is it possible to stack several Python extensions for the umachine module? Like one for Encoder/Counter, and another one for Pin?

robert-hh commented:
Just rebased to Version 1.19.xx

androiddrew commented on docs/mimxrt/quickref.rst:
```diff
@@ -537,6 +537,62 @@ port and LAN(1) for the 1G port.
+    qe.init(cpc=128)                # Specify 128 counts/cycle
+    qe.init(index=Pin(3))           # Specify Pin 3 as Index pulse input
+    qe.deinit()                     # turn off the Quadrature Encoder
+    qe.init(match=64)               # Set a match event at count 64
```
I flashed my teensy 4.1 with this branch's build. I am following this doc and can see values. When I attempt to use this match keyword I receive an error.

```
>>> e0.init(match=100)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: extra keyword arguments given
```

robert-hh replied:
Thank you for having a look at that class. It's always good if other people look at that. Here, the documentation has to be updated, since the match=xxx argument was removed from the init() call. That was done as result of trying to harmonize the various implementations. The same applies to the `reverse`keyword argument.

androiddrew commented on docs/mimxrt/quickref.rst:
```diff
@@ -537,6 +537,62 @@ port and LAN(1) for the 1G port.
+    qe = Encoder(0, Pin(0), Pin(1)) # create Quadrature Encoder object
+    qe.value()                      # get current counter values
+    qe.value(0)                     # Set value and cycles to 0
+    qe.init(cpc=128)                # Specify 128 counts/cycle
```
Counts per cycle(cpc) is a vague term for me. Should this be the Encoder's Counts Per Revolution(cpr), does it mean the Pulse Per Revolution(ppr)?

androiddrew commented on docs/library/machine.QECNT.rst:
```diff
@@ -0,0 +1,379 @@
+    qe.init(index=Pin(3))           # Specify Pin 3 as Index pulse input
+    qe.deinit()                     # turn off the Quadrature Encoder
+    qe.init(match=64)               # Set a match event at count 64
+    qe.irq(qe.IRQ_MATCH, value=100, handler=handler)
```
I am seeing errors with arguments also, but can successfully call `Encoder.irq` when using keyword arguments

```
>>> e0.irq(e0.IRQ_ROLL_OVER, handler=match_handler)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: extra positional arguments given
>>> e0.irq(trigger=e0.IRQ_ROLL_OVER, handler=match_handler)
```

robert-hh replied:
See above.

androiddrew commented on docs/library/machine.QECNT.rst:
```diff
@@ -0,0 +1,379 @@
+class Counter-- Signal counter for i.MXRT MCUs
+==============================================
+
+This class provides a Counter service using the Quadrature Encoder module
```
I think I have seen a couple of forum threads on the development approaches on these Encoder/Counter Classes, but as a new Micropython users, looking at the docs it is not clear why a second class exists and when to use one of these services over another.

robert-hh replied:
Encoder and Counter behave different. The Encoder requires at least in this implementation the two-phase signal, the counter works on a single phase.

robert-hh committed:
mimxrt: Fixes to the documentation of the Encoder/Counter class.

Thanks to @androiddrew for testing the code and finding the
inconsitencies to the documentation.

androiddrew commented:
I just encountered an issue when using my build from this branch. The `machine.UART` doesn't appear to have an `irq` method. I have a Bluetooth serial UART (ZS040) that I was going to fire a handler whenever data was received :( Is this a problem with this branch or does the Teensy UART not have support for an `irq`?

```
>>> UART.
any             read            readinto        readline
write           __bases__       __dict__        INV_RX
INV_TX          init            sendbreak
>>> UART.
```

robert-hh commented:
UART is interrupt based at the MIMXRT port. In the RX interrupt handler the receive buffer is filled. It is obviously possible to call a handler after the input queue has been drained. Maybe just with a handler=xxx keyword option.
UART.irq() exists at the moment only for the WiPy (CC3200) port, which is more or less legacy.

robert-hh commented:
The Teensy is pretty fast. So as a workaround you can set up a timer callback like every few ms and check for new data with uart.any().

robert-hh commented:
I looked into the code, and at least for MIMXRT it would require to rework the receive section. The codes uses a ringbuffer, but this buffer is part of the support library. It has a callback there, but that assumes that the target buffer for data is static. That is not the case for MicroPython. The other option would be to have the ringbuffer in the MP code part, like it's done for instance in the rpi 2040 port. Then full control over the behavior is possible.
But before that, one have to agree over the API. Using the existing API seems to be an overhead. It has many options with a single possible choice.

robert-hh committed:
mimxrt: Adapt for PR8813 - Optimise mp_obj_type_t storage.

IhorNehrutsa commented:
Hi, Robert.
May you please update the comparison table with latest PR
Proposal: Esp32 machine.PCNT through Encoder.py and Counter.py #8253
ESP32: Add Quadrature Encoder and Pulse Counter classes. #8766
esp32/esp32_pcnt: Add PCNT class and Counter/Encoder shims in machine #7582
Thanks.

robert-hh commented:
Hi @IhorNehrutsa I noticed the discussion about the various ESP32 Encoder PRs, but I did not look into the PRs except your's. And until no one of @dpgeorge or @jimmo shows and active feedback to the Encoder/Counter classes, this topic is on halt. I keep my PR mergeable, but that's it. The SAMD controller have Encoder/Counter hardware too, but without a general approach there is no sense in supporting it.

robert-hh committed:
mimxrt: Support Encoder/Counter for the MIMXRT117x family.

robert-hh commented:
Reducing the PR-essure by an epsilon.

androiddrew commented:
Sorry haven't been following this. Does this mean no encoder support for the teensy 4.x?

robert-hh commented:
The topic of Encoder has been discussed a lot, but besides that there is no progress. So I close it for now. I can always reopen or re-submit it. The branch with the encoder still exists at https://github.com/robert-hh/micropython/tree/mimxrt_qecnt and I will maintain it. Besides that you can always use the python implementations for encoders, as discussed in the PR message. At Teensy the work well up to a pulse rate of several kHz.

robert-hh commented:
Re-submitted as #12347
